src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,219 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import linear_algebra.matrix.symmetric
+import data.polynomial.monomial
+import data.matrix.pequiv
+import data.equiv.fin
+
+/-!
+# Circulant matrices
+
+This file contains the definition and basic results about circulant matrices.
+Given a vector `v : I → α` indexed by a type that is endowed with subtraction,
+`circulant v` is the matrix whose `(i,j)`th entry is `v (i = j)`.
+
+## Main results
+
+- `matrix.circulant`: the circulant matrix generated by a given vector `v : I → α`.
+
+## Implementation notes
+
+`matrix.fin.foo` is the `fin n` version of `matrix.foo`.
+Namely, the index type of the circulant matrices in discussion is `fin n`.
+
+## Tags
+
+circulant, matrix
+-/
+
+variables {α β I J R : Type*} {n : ℕ}
+
+namespace matrix
+
+open function
+open_locale matrix big_operators
+
+/-- Given the condition `[has_sub I]` and a vector `v : I → α`,
+    we define `circulant v` to be the circulant matrix generated by `v` of type `matrix I I α`.
+    The `(i,j)`th entry is defined to be `v (i - j)`. -/
+@[simp]
+def circulant [has_sub I] (v : I → α) : matrix I I α
+| i j := v (i - j)
+
+lemma circulant_col_zero_eq [add_group I] (v : I → α) :
+  (λ i, (circulant v) i 0) = v :=
+by ext; simp
+
+lemma circulant_injective [add_group I] : injective (circulant : (I → α) → matrix I I α) :=
+begin
+  intros v w h,
+  rw [← circulant_col_zero_eq v, ← circulant_col_zero_eq w, h]
+end
```

Comment:
Having `λ` at the head of the goal is pretty weird. I think this would be better as:
```suggestion
lemma circulant_col_zero_eq [add_group I] (v : I → α) (i : I) : circulant v i 0 = v i :=
congr_arg v (sub_zero _)

lemma circulant_injective [add_group I] : injective (circulant : (I → α) → matrix I I α) :=
begin
  intros v w h,
  ext k,
  rw [← circulant_col_zero_eq v, ← circulant_col_zero_eq w, h]
end
```